### DISCUSSION TITLE
NumericUpDown binding Value to converter

### DISCUSSION THREAD

**Discussion by Arcnor** [AvaloniaUI/Avalonia#16448] [Q&A] [answered]
I want to use a `NumericUpDown` and bind its `Value` to a `double` property (on code behind). `ValueProperty` is `decimal?` so I need a `Converter`. However, there is only `NumericUpDown.TextConverter` for the `Text` property. On WPF I can set the `Converter` property on the `Binding` itself, but it doesn't seem I can do that on Avalonia, unless I'm missing something?

Thanks in advance!

**Comment by IanRawley** [comment]
Have a read of these, they should help you out.
https://docs.avaloniaui.net/docs/basics/data/data-binding/data-binding-syntax#data-binding-markupextension
https://docs.avaloniaui.net/docs/guides/data-binding/how-to-create-a-custom-data-binding-converter

**Reply by Arcnor** [reply]
Hey @IanRawley , thanks for the answer! I've read those docs a few times already, but they deal with doing everything in XAML while I want to do it in code. Maybe there is an obvious conversion between the two, but I don't know how it can be.

Also to be clear, I can make it work one-way because most properties have a "converter" lambda when using `GetObservable` for example, but I need this to work two-way. I'm just missing something that might be ReactiveUI/CommunityMVVM related instead of Avalonia, but still, I'm really confused as to how to make this work :).

Thanks again, appreciate it!

**Reply by IanRawley** [reply, marked as answer]
I did start wondering if that was what you meant by "on code behind".  I'd originally taken it to mean the property was on the view itself rather than viewmodel.

For doing it in code, take a look at this one: 
https://docs.avaloniaui.net/docs/guides/data-binding/binding-from-code

Specifically any of the examples using XAML binding which refer to new Binding().  Binding has a ValueConverter property (inherited from BindingBase), taking IValueConverter? which while it's not used in the examples in that page should do what you need it to do.

**Reply by Arcnor** [reply]
Right, sorry, I thought "code behind" was the proper term, I've seen it used in documentation :).

In any case, I also saw the `Binding` class, but it only takes a string for the path, which I'm sure it's useful to reproduce those examples when everything lives on XAML, but I have nameless (as far as I know) controls that I create on the fly. Maybe I should be setting names & using `Binding` with the name of the new controls, but that seems really convoluted given I have a reference to both controls, I should be able to do the binding without any string parsing involved?

Not sure if that makes sense :).

**Reply by Arcnor** [reply]
Ok, I was obviously getting confused here, but it really works using `Binding`, the `path` is the path to the property, not to the control (that's the `Source` :D). Still, I'd love to be able to do this using a strong reference instead of a "weak" string parsing at runtime.

But I'll accept your answer, as it's the right way to do it anyway, thanks again!

**Reply by IanRawley** [reply]
That was me taking things weirdly.  I attached (on code behind) to the word property rather than bind, so went down the wrong track.  There probably is a better way to do code behind bindings that use ValueConverters, but that's not something I do so I'm not that familiar with it.  Good luck.
